### PR TITLE
refactor: builtin functions inherit from `VyperType`

### PR DESCRIPTION
**Pull request by tserg**
### What I did

Builtin functions are currently the remaining types that are not a "VyperType". This PR resolves this inconsistency by making builtin functions inherit from `VyperType`.

### How I did it

Move code around.

### How to verify it

See tests

### Commit message

```
chore: make builtin functions vyper types

Builtin functions are currently the remaining types that 
are not a "VyperType". This PR resolves this inconsistency 
by making builtin functions inherit from `VyperType`.
```

### Description for the changelog

Make builtin functions vyper types

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/semantics/types/function.py [hunk 1]
```diff
@@ -631,6 +636,79 @@ def fetch_call_return(self, node: vy_ast.Call) -> Optional[VyperType]:
...
+
+    def _validate_arg_types(self, node: vy_ast.Call) -> None:
+        num_args = len(self._inputs)  # the number of args the signature indicates
+
+        expect_num_args: Union[int, tuple] = num_args
+        if self._has_varargs:
+            # note special meaning for -1 in validate_call_args API
+            expect_num_args = (num_args, -1)
+
+        validate_call_args(node, expect_num_args, list(self._kwargs.keys()))
```
Comment (tserg) [inline vyper/semantics/types/function.py:664]: Last argument should be a list of valid keyword arguments. Previously, it was a dictionary of `KwargSettings`.

File: vyper/semantics/types/function.py [hunk 2]
```diff
@@ -631,6 +636,79 @@ def fetch_call_return(self, node: vy_ast.Call) -> Optional[VyperType]:
         return self.return_type
 
 
+class BuiltinFunctionT(VyperType):
+    _has_varargs = False
+    _inputs: list[tuple[str, Any]] = []
+    _kwargs: Dict[str, KwargSettings] = {}
+    _return_type: Optional[VyperType] = None
```
Comment (tserg) [inline vyper/semantics/types/function.py:643]: Added `_return_type` and `_inputs` attributes to make mypy happy.

File: vyper/semantics/types/function.py [hunk 3]
```diff
@@ -631,6 +636,79 @@ def fetch_call_return(self, node: vy_ast.Call) -> Optional[VyperType]:
...
+            # ensures the type can be inferred exactly.
+            get_exact_type_from_node(arg)
+
+    def fetch_call_return(self, node: vy_ast.Call) -> Optional[VyperType]:
+        self._validate_arg_types(node)
+
+        if self._return_type:
+            return self._return_type
+
+        return None
```
Comment (tserg) [inline vyper/semantics/types/function.py:691]: Added `None` return to make mypy happy.

File: vyper/semantics/types/function.py [hunk 4]
```diff
@@ -631,6 +636,79 @@ def fetch_call_return(self, node: vy_ast.Call) -> Optional[VyperType]:
...
+            assert self._has_varargs  # double check validate_call_args
+        for arg in varargs:
+            # call get_exact_type_from_node for its side effects -
+            # ensures the type can be inferred exactly.
+            get_exact_type_from_node(arg)
+
+    def fetch_call_return(self, node: vy_ast.Call) -> Optional[VyperType]:
+        self._validate_arg_types(node)
+
+        if self._return_type:
```
Comment (charles-cooper) [inline vyper/semantics/types/function.py:688]: i think just `return self._return_type`?

File: vyper/builtins/functions.py [hunk 5]
```diff
@@ -74,6 +74,7 @@
     TupleT,
 )
 from vyper.semantics.types.bytestrings import _BytestringT
+from vyper.semantics.types.function import BuiltinFunctionT
```
[the remark this answers is not in the document]
Comment (charles-cooper) [reply vyper/builtins/functions.py:77]: import cycle: vyper.semantics.types.function -> vyper.semantics.analysis.utils -> vyper.semantics.namespace -> (inside of Namespace ctor) vyper.builtins.functions

File: vyper/builtins/_signatures.py [hunk 6]
```diff
@@ -89,15 +91,15 @@ def _validate_single(self, arg, expected_type):
         else:
             validate_expected_type(arg, expected_type)
 
-    def _validate_arg_types(self, node):
+    def _validate_arg_types(self, node: vy_ast.Call) -> None:
         num_args = len(self._inputs)  # the number of args the signature indicates
 
-        expect_num_args = num_args
+        expect_num_args: Union[int, tuple] = num_args
```
Comment (charles-cooper) [inline vyper/builtins/_signatures.py:97]: why is this a union of int and tuple?

**Review by charles-cooper — APPROVED** [review]

**Pull request merged**